File: packages/expo-dev-menu/vendored/react-native-safe-area-context/ios/SafeAreaView/RCTView+DevMenuSafeAreaCompat.m
#import "RCTView+DevMenuSafeAreaCompat.h"

#import <React/RCTUIManager.h>

BOOL DevMenuUIEdgeInsetsEqualToEdgeInsetsWithThreshold(UIEdgeInsets insets1, UIEdgeInsets insets2, CGFloat threshold)
{
  return ABS(insets1.left - insets2.left) <= threshold &&
    ABS(insets1.right - insets2.right) <= threshold &&
    ABS(insets1.top - insets2.top) <= threshold &&
    ABS(insets1.bottom - insets2.bottom) <= threshold;
}

@implementation UIView(DevMenuSafeAreaCompat)

- (BOOL)nativeSafeAreaSupport
{
  return [self respondsToSelector:@selector(safeAreaInsets)];
}

- (UIEdgeInsets)safeAreaInsetsOrEmulate
{
  #if defined(__IPHONE_OS_VERSION_MAX_ALLOWED) && __IPHONE_OS_VERSION_MAX_ALLOWED >= 110000 /* __IPHONE_11_0 */
  if (self.nativeSafeAreaSupport) {
    if (@available(iOS 11.0, *)) {
      return self.safeAreaInsets;
    }
  }
  #endif
  return self.emulatedSafeAreaInsets;
}

- (UIEdgeInsets)emulatedSafeAreaInsets
{
  UIViewController* vc = self.reactViewController;

  if (!vc) {
    return UIEdgeInsetsZero;
  }

  CGFloat topLayoutOffset = vc.topLayoutGuide.length;
  CGFloat bottomLayoutOffset = vc.bottomLayoutGuide.length;
  CGRect safeArea = vc.view.bounds;
  safeArea.origin.y += topLayoutOffset;
  safeArea.size.height -= topLayoutOffset + bottomLayoutOffset;
  CGRect localSafeArea = [vc.view convertRect:safeArea toView:self];
  UIEdgeInsets safeAreaInsets = UIEdgeInsetsMake(0, 0, 0, 0);
  if (CGRectGetMinY(localSafeArea) > CGRectGetMinY(self.bounds)) {
    safeAreaInsets.top = CGRectGetMinY(localSafeArea) - CGRectGetMinY(self.bounds);
  }
  if (CGRectGetMaxY(localSafeArea) < CGRectGetMaxY(self.bounds)) {
    safeAreaInsets.bottom = CGRectGetMaxY(self.bounds) - CGRectGetMaxY(localSafeArea);
  }

  return safeAreaInsets;
}

@end

File: packages/expo-dev-menu/vendored/react-native-safe-area-context/ios/SafeAreaView/RCTView+DevMenuSafeAreaCompat.swift
import React

func DevMenuUIEdgeInsetsEqualToEdgeInsetsWithThreshold(_ insets1: UIEdgeInsets, _ insets2: UIEdgeInsets, _ threshold: CGFloat) -> Bool {
  return abs(insets1.left - insets2.left) <= threshold &&
    abs(insets1.right - insets2.right) <= threshold &&
    abs(insets1.top - insets2.top) <= threshold &&
    abs(insets1.bottom - insets2.bottom) <= threshold
}

extension UIView {
  var nativeSafeAreaSupport: Bool {
    return responds(to: #selector(getter: safeAreaInsets))
  }

  func safeAreaInsetsOrEmulate() -> UIEdgeInsets {
    if nativeSafeAreaSupport {
      if #available(iOS 11.0, *) {
        return safeAreaInsets
      }
    }
    return emulatedSafeAreaInsets()
  }

  func emulatedSafeAreaInsets() -> UIEdgeInsets {
    guard let vc = self.reactViewController() else {
      return .zero
    }

    let topLayoutOffset = vc.topLayoutGuide.length
    let bottomLayoutOffset = vc.bottomLayoutGuide.length
    var safeArea = vc.view.bounds
    safeArea.origin.y += topLayoutOffset
    safeArea.size.height -= topLayoutOffset + bottomLayoutOffset
    let localSafeArea = vc.view.convert(safeArea, to: self)
    var safeAreaInsets = UIEdgeInsets.zero

    if localSafeArea.minY > bounds.minY {
      safeAreaInsets.top = localSafeArea.minY - bounds.minY
    }

    if localSafeArea.maxY < bounds.maxY {
      safeAreaInsets.bottom = bounds.maxY - localSafeArea.maxY
    }

    return safeAreaInsets
  }
}
